### PR TITLE
Shoppincart

### PR DESCRIPTION
**Pull request by Oleksandr-Radovets**

### CODE REVIEW

File: src/main/java/com/example/securitynew/model/ShoppingCart.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shoppingCart")
+public class ShoppingCart {
+    @Id
+    @Column(name = "shoppingCart_id")
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long idShoppingCart;
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/ShoppingCart.java]: primary keys should be named like that
```suggestion
    private Long id;
```

File: src/main/java/com/example/securitynew/model/ShoppingCart.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shoppingCart")
+public class ShoppingCart {
+    @Id
+    @Column(name = "shoppingCart_id")
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/ShoppingCart.java]: this changes nothing
```suggestion
```

File: src/main/java/com/example/securitynew/model/ShoppingCart.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import jakarta.persistence.Table;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shoppingCart")
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/ShoppingCart.java]: name should be default (default name in db will be 'shopping_cart')
```suggestion
```

File: src/main/java/com/example/securitynew/model/ShoppingCart.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long idShoppingCart;
+    @OneToOne
+    @JoinTable(name = "shoppingCart_users", joinColumns = @JoinColumn(name = "shoppingCart_id"),
+            inverseJoinColumns = @JoinColumn(name = "user_id"))
+    private User user;
+    @OneToMany(cascade = CascadeType.REMOVE, fetch = FetchType.EAGER)
+    @JoinTable(name = "shoppingCart_cartItems", joinColumns = @JoinColumn(name = "shoppingCart_id"),
+            inverseJoinColumns = @JoinColumn(name = "cartItems_id"))
+    private Set<CartItem> cartItemSet = new HashSet<>();
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/ShoppingCart.java]:
```suggestion
    private Set<CartItem> cartItems = new HashSet<>();
```

File: src/main/java/com/example/securitynew/repository/CartItemRepository.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.example.securitynew.repository;
+
+import com.example.securitynew.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+
+
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/repository/CartItemRepository.java]:
```suggestion
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+@Getter
+@Setter
+@Entity
+@Table(name = "cartItem")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "cart_item_id")
+    private Long idCartItem;
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/CartItem.java]: same
```suggestion
    private Long id;
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.JoinTable;
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "cartItem")
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/CartItem.java]: same
```suggestion
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+            joinColumns = @JoinColumn(name = "cart_item_id"),
+            inverseJoinColumns = @JoinColumn(name = "shopping_cart_id"))
+    private ShoppingCart shoppingCart;
+    @OneToOne
+    @JoinTable(name = "cartItem_books",
+            joinColumns = @JoinColumn(name = "cart_item_id"),
+            inverseJoinColumns = @JoinColumn(name = "book_id"))
+    private Book book;
+    private int quantity;
+
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/model/CartItem.java:39]:
```suggestion
```

File: src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> creatShoppingCart(user));
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(bookRepository.findById(idBook).get());
+        cartItem.setQuantity(quantity);
+        shoppingCart.getCartItemSet().add(cartItemRepository.save(cartItem));
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    private ShoppingCart creatShoppingCart(User user) {
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    private ShoppingCart createShoppingCart(User user) {
```

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/securitynew/service/ShoppingCartService.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package com.example.securitynew.service;
+
+import com.example.securitynew.dto.shoppingcart.ShoppingCartResponseDto;
+
+public interface ShoppingCartService {
+
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/service/ShoppingCartService.java:6]: In CategoryService you have no empty line and here is, keep same style

File: src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private ShoppingCart createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        return shoppingCart;
+    }
+
+    public ShoppingCartResponseDto getShoppingCartUser(Long id) {
+        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(id).get());
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java:54]: Use just get method from optional is not good idea

File: src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public ShoppingCartResponseDto update(Long idCartItem, int quantity) {
+        String name = SecurityContextHolder.getContext().getAuthentication().getName();
+        User user = userRepository.findByEmail(name).get();
+        ShoppingCart byUserId = shoppingCartRepository.findByUserId(user.getId()).get();
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java]: Why do you get user form repository instead of service layer?

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/securitynew/controller/ShoppingCartController.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api")
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/controller/ShoppingCartController.java]:
```suggestion
@RequestMapping("/api/cart")
```

File: src/main/java/com/example/securitynew/dto/cartitem/CartItemRequestDto.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package com.example.securitynew.dto.cartitem;
+
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    private Long idCartItem;
+    private Long idBook;
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/dto/cartitem/CartItemRequestDto.java]: Add validation
```suggestion
    private Long cartItemId;
    private Long bookId;
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import jakarta.persistence.JoinTable;
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "cartItem")
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/model/CartItem.java]:
```suggestion
@Table(name = "cart_items")
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinTable(name = "cartItem_shoppingCart",
+            joinColumns = @JoinColumn(name = "cart_item_id"),
+            inverseJoinColumns = @JoinColumn(name = "shopping_cart_id"))
+    private ShoppingCart shoppingCart;
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/model/CartItem.java]: OneToOne?

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    @OneToOne
+    @JoinTable(name = "cartItem_books",
+            joinColumns = @JoinColumn(name = "cart_item_id"),
+            inverseJoinColumns = @JoinColumn(name = "book_id"))
+    private Book book;
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/model/CartItem.java]: OneToOne?

File: src/main/java/com/example/securitynew/model/ShoppingCart.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.Table;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shopping_cart")
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/model/ShoppingCart.java]:
```suggestion
@Table(name = "shopping_carts")
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "cartItem")
+public class CartItem {
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/model/CartItem.java:22]: Use soft delete

File: src/main/java/com/example/securitynew/model/ShoppingCart.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+public class ShoppingCart {
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/model/ShoppingCart.java:27]: Use soft delete

File: src/main/java/com/example/securitynew/service/impl/BookServiceImpl.java [hunk 21]
```diff
@@ -48,11 +48,19 @@ public void delete(Long id) {
...
+    public List<BookDto> findAllByCategoryId(Long categoryId) {
+        return bookRepository.findAllByCategoryId(categoryId)
+                .stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
+
     @Override
     public void updateById(Long id, CreateBookRequestDto createBookRequestDto) {
         Book book = bookRepository.findById(id)
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/service/impl/BookServiceImpl.java:61]: do not find Book - just check id

File: src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java (hunk 11 above)
Comment (Rommelua) [reply src/main/java/com/example/securitynew/service/impl/ShoppingCartServiceImpl.java:54]: not fixed

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by koval-kn — APPROVED** [review]

**Pull request merged**